### ISSUE TITLE
dplyr / spark_read_csv() / High Column count Bug

### ISSUE THREAD

**Issue by yarrick19** [rstudio/sparklyr#300]
@javierluraschi @kevinushey 
Hey guys,

I ran into a bug using a combination of `dplyr` on a spark df that was created with `spark_read_csv()` that has more than 200 columns.  Since the built in `sdf_partition()` method has issues pertaining to large column counts ( issue #264 ) I'm attempting to work around this by appending a column to my dataset that contains a random number and then using that to partition the data.

There seems to be a bug that occurs when referencing the random number column which causes the rest of the data in the data frame to not be correctly reflected.  Below is a reproducible example.  What you'll see is that the v1 variable values (and all others if you look at the data) all change to 0 after partitioning based on the new random number column.  

```
library(sparklyr)
library(dplyr)
sc <- spark_connect(master = "local")


# --------------------------------------------
# Simulate Dataset
# --------------------------------------------

df <- as.data.frame(matrix(0:1, ncol = 202, nrow = 100))
# sdf <- copy_to(sc, df, "df_test", overwrite = TRUE)

 
# export/import to CSV
file_name <- "C:\\Users\\pnieman\\Documents\\R\\lookaliketest\\data\\sdfrandtest.csv"
write.csv(x = df,file = file_name,row.names = FALSE)
sdf <- spark_read_csv(sc, "rt", path = file_name, overwrite = TRUE)


# Look at counts by v1 variable (You can see that v1 has both 1s and 0s)
sdf %>%
    group_by(v1) %>%
    summarise(ct = n())


# --------------------------------------------
# Partition Data
# --------------------------------------------

randtest <- sdf %>%
    mutate(rannum = rand()) %>%
    sdf_register("rt_diw")

# Partition based on Random Number 

partition_1 <- randtest %>%
    filter(rannum <= 0.8)

partition_2 <- randtest %>%
    filter(rannum > 0.8)


# Look at counts by "target" variable based on partitions (now everything are 0s!)
partition_1 %>%
    group_by(v1) %>%
    summarise(ct = n())

partition_2 %>%
    group_by(v1) %>%
    summarise(ct = n())
```

Other observations:

- This only occurs when there are more than 200 columns.  You can change this example to see.

- When using the `copy_to()` method instead of `spark_read_csv()` this seems to work as expected.  This code is included - just need to uncomment it out for testing and comment out the CSV section.  This leads me to believe that there may be some underlying issues with `spark_read_csv()`.

- The data in the `rt_diw` spark df looks correct, yet subsequent queries which use the rannum column result in bad data.

- No error is thrown in this... yet the data changes.  A little scary.

**Comment by kevinushey** [comment]
What version of Spark are you using? This is, I believe, a (very terrifying!) bug in Spark 2.0.0 and below. However, I can't reproduce with Spark 2.0.1.

**Comment by yarrick19** [comment]
@kevinushey 
I was using Spark 2.0.0.  I've installed 2.0.1 and it seems to work as expected.  Thanks for the suggestion!  Hopefully they back port those fixes into previous versions of Spark - seems like a major issue.  Should I close this issue?

**Comment by kevinushey** [comment]
Indeed, I think this can be closed. For reference, the bug is tracked upstream at https://issues.apache.org/jira/browse/SPARK-16664; it sounds like a fix will be part of Spark 1.6.3 as well.

**Comment by dsblr** [comment]
@kevinushey I am trying to get the number of rows and number of columns after reading the file from csv. But I am unable to get the number of rows. Please suggest some solution.
test_data <- spark_read_csv(sc = sc, name = "test1",path ="C:/Users/user1/data/bk_test1.csv")

> dim(test_data)
[1] NA 15
> ncol(test_data)
[1] 15
> nrow(test_data)
[1] NA
